netket/graph/graph.py

```diff
@@ -103,19 +103,18 @@ def automorphisms(self) -> PermutationGroup:
                 "automorphisms is not yet implemented for colored edges"
             )
 
-        if self._automorphisms is not None:
-            return self._automorphisms
+        if self.__automorphisms is not None:
+            return self.__automorphisms
```

Comment:
This logic was there in the original `NetworkX` class. It makes sense to make them immutable, but you do need a nontrivial constructor for them...

Reply:
I don't think graphs were explicitly immutable, though I've never felt the need to change one. I think we can make it explicitly immutable.

Reply:
Maybe it's good to spell it out if we assume it. Also it means that we can use `struct.dataclass` rather than suffering with custom implementations of `struct.property_cached`. @PhilipVinc could you explain how `pre_init` works? the constructor here contains a fair bit of custom logic

Reply:
Sorry I just noticed this.
If you look at the docstring on `netket/utils/dataclass` there is an example showing what it does exactly. In short, `__pre_init__` is run before the data class init (much like `_pos _init_` is run after).
`pre_init` therefore defines what arguments/keywordargs are accepted when you construct the object, and it should return the args/keyword args accepted/needed by the data class.

You can think of it as a function that can take any argument, and those arguments will then be those accepted by the constructor of the object.
It should return a tuple where the first element is an empty tuple and the second is a dictionary with a value for every field in the data class.

Reply:
Basically, the data class init

```
def __init__(*args, **kwargs):
   #autogenerated by data class 
```
becomes
```
def __init__(*args, **kwargs):
   args, kwargs = self.__pre_init__(*args, **kwargs)
   self.__dataclass_init__(*args, **kwargs)

def __dataclass_init__(*args, **kwargs):
   #autogenerated by data class 

```